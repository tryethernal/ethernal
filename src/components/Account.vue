<template>
    <div>
        <v-row>
            <v-col cols="6">
                <h4>Api Token</h4>
                <v-card outlined class="mb-4">
                    <v-card-text v-if="apiToken && !loading">
                        <v-text-field id="apiToken" append-icon="mdi-content-copy" readonly @click:append="copyToken()" outlined dense hide-details="auto" :value="apiToken" v-show="apiToken"></v-text-field>
                        <input type="hidden" id="copyElement" :value="apiToken">
                    </v-card-text>
                    <v-card-text v-else>
                        <v-skeleton-loader type="list-item-two-line"></v-skeleton-loader>
                    </v-card-text>
                </v-card>
            </v-col>
        </v-row>
    </div>
</template>
<script>

export default {
    name: 'Account',
    data: () => ({
        apiToken: null,
        loading: false
    }),
    mounted: function() {
        this.loading = true;
        this.server.getApiToken()
            .then(({ data: { apiToken }}) => this.apiToken = apiToken)
            .catch(console.log)
            .finally(() => this.loading = false);
    },
    methods: {
        copyToken: function() {
            const webhookField = document.querySelector('#copyElement');
            webhookField.setAttribute('type', 'text');
            webhookField.select();

            try {
                const copied = document.execCommand('copy');
                const message = copied ? 'Token copied!' : `Couldn't copy token`;
                alert(message);
            } catch(error) {
                alert(`Couldn't copy token`);
            } finally {
                webhookField.setAttribute('type', 'hidden');
                window.getSelection().removeAllRanges();
            }
        },
    }
}
</script>
