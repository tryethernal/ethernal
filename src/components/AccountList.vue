<template>
    <v-container fluid>
        <h2 class="text-h6 font-weight-medium">Top Accounts by {{ explorerStore.token }} Balance</h2>
        <v-divider class="my-4"></v-divider>
        <v-card class="mt-4">
            <v-card-text>
                <v-data-table-server
                    class="hide-table-count"
                    :dense="dense"
                    :loading="loading"
                    :items="accounts"
                    :items-length="accountCount"
                    :sort-by="currentOptions.sortBy"
                    :must-sort="true"
                    :headers="headers"
                    :hide-default-footer="dense"
                    :hide-default-header="dense"
                    no-data-text="No accounts indexed yet"
                    last-icon=""
                    first-icon=""
                    :items-per-page-options="[
                        { value: 10, title: '10' },
                        { value: 25, title: '25' },
                        { value: 100, title: '100' }
                    ]"
                    item-key="address"
                    @update:options="getAccounts">
                    <template #item.address="{ item }">
                        <Hash-Link :contract="item.contract" :fullHash="true" :type="'address'" :hash="item.address" />
                    </template>
                    <template #item.balance="{ item }">
                        {{ fromWei(item.balance, 18, explorerStore.token) }}
                    </template>
                    <template #item.share="{ item }">
                        {{ (item.share * 100).toFixed(2) }}%
                    </template>
                    <template #item.transaction_count="{ item }">
                        {{ item.transaction_count }}
                    </template>
                    <template v-slot:loading>
                        <div class="d-flex justify-center align-center pa-4">
                            <v-progress-circular
                                size="24"
                                width="2"
                                indeterminate
                                color="primary"
                                class="mr-3"
                            ></v-progress-circular>
                            Loading balances...
                        </div>
                    </template>
                </v-data-table-server>
            </v-card-text>
        </v-card>
    </v-container>
</template>

<script setup>
import { ref, inject, onMounted } from 'vue';
import { useExplorerStore } from '../stores/explorer';
import HashLink from './HashLink.vue';
import fromWei from '../filters/FromWei';

const $server = inject('$server');

const explorerStore = useExplorerStore();

const dense = ref(false);
const accounts = ref([]);
const accountCount = ref(0);
const loading = ref(false);
const currentOptions = ref({ page: 1, itemsPerPage: 10, orderBy: 'address', order: 'desc' });
const headers = ref([
    { title: 'Address', key: 'address', sortable: false },
    { title: 'Balance', key: 'balance', sortable: false },
    { title: 'Percentage', key: 'share', sortable: false },
    { title: 'Transaction Count', key: 'transaction_count', sortable: false }
]);


function getAccounts({ page, itemsPerPage, sortBy } = {}) {
    if (loading.value)
        return;

    if (!page || !itemsPerPage || !sortBy || !sortBy.length)
        return loading.value = false;
    
    currentOptions.value = { page, itemsPerPage, sortBy };

    $server.getWorkspaceFilteredNativeAccounts({ page, itemsPerPage })
        .then(({ data }) => {
            if (data.total)
                accountCount.value = data.total;
            else
                accountCount.value = data.items.length == currentOptions.value.itemsPerPage ?
                    (currentOptions.value.page * data.items.length) + 1 :
                    currentOptions.value.page * data.items.length;

            accounts.value = data.items;
        })
        .catch(console.log)
        .finally(() => loading.value = false);
}

onMounted(() => {
    currentOptions.value.sortBy = [{ key: 'balance', order: 'desc' }];
});
</script>
