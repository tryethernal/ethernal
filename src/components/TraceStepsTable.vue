<template>
    <v-data-table-server
        class="hide-table-count"
        :loading="loading"
        :items="items"
        :items-length="totalItems"
        :sort-by="sortBy"
        :must-sort="true"
        items-per-page-text="Rows per page:"
        :headers="headers"
        :hide-default-footer="dense"
        :hide-default-header="dense"
        no-data-text="No internal transactions indexed yet"
        last-icon=""
        first-icon=""
        :items-per-page-options="[
            { value: 10, title: '10' },
            { value: 25, title: '25' },
            { value: 100, title: '100' }
        ]"
        item-key="id"
        @update:options="onUpdateOptions">
        <template v-slot:footer.page-text>
        </template>
        
        <template v-slot:item.transaction="{ item }">
            <Hash-Link :type="'transaction'" :hash="item.transaction.hash" :xsHash="true" />
        </template>

        <template v-slot:item.method="{ item }">
            <v-tooltip v-if="item.method.name">
                <template v-slot:activator="{ props }">
                    <v-chip v-bind="props" color="primary" label size="small" variant="tonal">
                        {{ item.method.name }}
                    </v-chip>
                </template>
                <span style="white-space: pre">{{ item.method.label }}</span>
            </v-tooltip>
            <v-chip v-else-if="item.method.sighash" color="primary" label size="small" variant="tonal">
                {{ item.method.sighash }}
            </v-chip>
        </template>
        
        <template v-slot:item.timestamp="{ item }">
            <div class="my-2 text-left">
                {{ $dt.shortDate(item.transaction.timestamp) }}<br>
                <small class="text-caption text-medium-emphasis">{{ $dt.fromNow(item.transaction.timestamp) }}</small>
            </div>
        </template>
        
        <template v-slot:item.from="{ item }">
            <template v-if="dense">
                <div class="my-2 text-left">
                    From: <Hash-Link :type="'address'" :hash="item.from.address" /><br>
                </div>
            </template>
            <template v-else>
                <v-chip size="x-small" class="mr-2" v-if="item.from && item.from.address === highlightAddress">self</v-chip>
                <Hash-Link :contract="item.from.address != highlightAddress ? item.from.contract : null" :type="'address'" :hash="item.from.address" :withTokenName="true" :withName="true" />
            </template>
        </template>

        <template v-slot:item.to="{ item }">
            <v-chip size="x-small" class="mr-2" v-if="item.to && item.to.address === highlightAddress">self</v-chip>
            <Hash-Link :contract="item.to.address != highlightAddress ? item.to.contract : null" :type="'address'" :hash="item.to.address" :withTokenName="true" :withName="true" />
        </template>
        
        <template v-slot:item.value="{ item }">
            {{ $fromWei(item.value, 'ether', currentWorkspaceStore.chain.token, false, 4) }}
        </template>
        
        <template v-slot:item.type="{ item }">
            <v-chip label size="small" color="primary" variant="tonal">{{ item.op }}</v-chip>
        </template>
    </v-data-table-server>
</template>

<script setup>
import { inject } from 'vue';
import { useCurrentWorkspaceStore } from '../stores/currentWorkspace';
import HashLink from './HashLink.vue';

const props = defineProps({
    dense: Boolean,
    highlightAddress: {
        type: String,
        default: ''
    },
    loading: {
        type: Boolean,
        required: true
    },
    items: {
        type: Array,
        required: true
    },
    totalItems: {
        type: Number,
        required: true
    },
    sortBy: {
        type: Array,
        default: () => [{ key: 'timestamp', order: 'desc' }]
    }
});

const emit = defineEmits(['update:options']);

const currentWorkspaceStore = useCurrentWorkspaceStore();
const $dt = inject('$dt');
const $fromWei = inject('$fromWei');

const headers = [
    { title: 'Transaction', key: 'transaction', align: 'start', sortable: false },
    { title: 'Type', key: 'type', sortable: false },
    { title: 'Method', key: 'method', sortable: false },
    { title: 'Mined On', key: 'timestamp', sortable: false },
    { title: 'From', key: 'from', sortable: false },
    { title: 'To', key: 'to', sortable: false },
    { title: 'Value', key: 'value', sortable: false }
];

const onUpdateOptions = (options) => {
    emit('update:options', options);
};

</script>
