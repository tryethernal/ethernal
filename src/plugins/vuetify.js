import Vue from 'vue';
import Vuetify from 'vuetify/lib/framework';

Vue.use(Vuetify);

export default new Vuetify({
    theme: {
        options: {
            customProperties: true
        },
        themes: {
            light: {
                primary: '#3D95CE',
                secondary: '#424242',
                accent: '#29648E',
                error: '#E72732',
                info: '#2196F3',
                success: '#4CAF50',
                warning: '#E78227',
                background: '#F5F5F5'
            }
        }
    }
});
