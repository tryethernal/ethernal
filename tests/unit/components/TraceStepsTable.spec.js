import TraceStepsTable from '@/components/TraceStepsTable.vue';
import { VDataTableServer } from 'vuetify/components';

const stubs = ['Hash-Link'];

describe('TraceStepsTable.vue', () => {
    const mockItems = [{
        id: '1',
        transaction: {
            hash: '0x123',
            timestamp: '2024-03-21T10:00:00Z'
        },
        method: {
            name: 'transfer',
            label: 'Transfer(address,uint256)',
            sighash: '0xa9059cbb'
        },
        from: {
            address: '0xabc',
            contract: null
        },
        to: {
            address: '0xdef',
            contract: {
                name: 'TestToken'
            }
        },
        value: '1000000000000000000',
        op: 'CALL'
    }];

    const defaultProps = {
        loading: false,
        items: mockItems,
        totalItems: 1,
        sortBy: [{ key: 'timestamp', order: 'desc' }]
    };

    it('Should render the table with data', async () => {
        const wrapper = mount(TraceStepsTable, {
            props: defaultProps,
            global: {
                stubs,
                provide: {
                    $dt: {
                        shortDate: vi.fn().mockReturnValue('Mar 21, 2024'),
                        fromNow: vi.fn().mockReturnValue('2 hours ago')
                    },
                    $fromWei: vi.fn().mockReturnValue('1.0')
                }
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should render the table in dense mode', async () => {
        const wrapper = mount(TraceStepsTable, {
            props: {
                ...defaultProps,
                dense: true
            },
            global: {
                stubs,
                provide: {
                    $dt: {
                        shortDate: vi.fn().mockReturnValue('Mar 21, 2024'),
                        fromNow: vi.fn().mockReturnValue('2 hours ago')
                    },
                    $fromWei: vi.fn().mockReturnValue('1.0')
                }
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should render the table with highlighted address', async () => {
        const wrapper = mount(TraceStepsTable, {
            props: {
                ...defaultProps,
                highlightAddress: '0xabc'
            },
            global: {
                stubs,
                provide: {
                    $dt: {
                        shortDate: vi.fn().mockReturnValue('Mar 21, 2024'),
                        fromNow: vi.fn().mockReturnValue('2 hours ago')
                    },
                    $fromWei: vi.fn().mockReturnValue('1.0')
                }
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should emit update:options when table options change', async () => {
        const wrapper = mount(TraceStepsTable, {
            props: defaultProps,
            global: {
                stubs,
                provide: {
                    $dt: {
                        shortDate: vi.fn().mockReturnValue('Mar 21, 2024'),
                        fromNow: vi.fn().mockReturnValue('2 hours ago')
                    },
                    $fromWei: vi.fn().mockReturnValue('1.0')
                }
            }
        });

        const table = wrapper.findComponent(VDataTableServer);
        const newOptions = {
            page: 1,
            itemsPerPage: 10,
            sortBy: [{ key: 'timestamp', order: 'desc' }],
            groupBy: [],
            search: undefined
        };
        await table.vm.$emit('update:options', newOptions);

        expect(wrapper.emitted('update:options')).toBeTruthy();
        expect(wrapper.emitted('update:options')[0]).toEqual([newOptions]);
    });
}); 