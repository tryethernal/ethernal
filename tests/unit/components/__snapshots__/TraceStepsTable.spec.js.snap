// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TraceStepsTable.vue > Should render the table in dense mode 1`] = `
"<div class="v-table v-table--has-top v-table--has-bottom v-theme--light v-table--density-default v-data-table hide-table-count" item-key="id">
  <!---->
  <div class="v-table__wrapper">
    <table>
      <!---->
      <!---->
      <!---->
      <tbody class="v-data-table__tbody" role="rowgroup">
        <!---->
        <tr class="v-data-table__tr">
          <td class="v-data-table__td v-data-table-column--align-start">
            <hash-link-stub type="transaction" hash="0x123" fullhash="false" withname="true" notcopiable="false" withtokenname="false" xshash="true" unlink="false" loadcontract="false"></hash-link-stub>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start"><span class="v-chip v-chip--label v-theme--light text-primary v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false"><!----><span class="v-chip__underlay"></span>
            <!---->
            <!---->
            <div class="v-chip__content" data-no-activator="">CALL</div>
            <!---->
            <!----></span>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start"><span class="v-chip v-chip--label v-theme--light text-primary v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false" aria-describedby="v-tooltip-1"><!----><span class="v-chip__underlay"></span>
            <!---->
            <!---->
            <div class="v-chip__content" data-no-activator="">transfer</div>
            <!---->
            <!----></span>
            <!--teleport start-->
            <!--teleport end-->
          </td>
          <td class="v-data-table__td v-data-table-column--align-start">
            <div class="my-2 text-left">03/21 11:00:00 AM<br><small class="text-caption text-medium-emphasis">in 1 year</small></div>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start">
            <div class="my-2 text-left"> From: <hash-link-stub type="address" hash="0xabc" fullhash="false" withname="true" notcopiable="false" withtokenname="false" xshash="false" unlink="false" loadcontract="false"></hash-link-stub><br></div>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start">
            <!--v-if-->
            <hash-link-stub contract="[object Object]" type="address" hash="0xdef" fullhash="false" withname="true" notcopiable="false" withtokenname="true" xshash="false" unlink="false" loadcontract="false"></hash-link-stub>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start">1 ETH</td>
        </tr>
        <!---->
        <!---->
      </tbody>
      <!---->
      <!---->
    </table>
  </div>
  <!---->
</div>"
`;

exports[`TraceStepsTable.vue > Should render the table with data 1`] = `
"<div class="v-table v-table--has-top v-table--has-bottom v-theme--light v-table--density-default v-data-table hide-table-count" item-key="id">
  <!---->
  <div class="v-table__wrapper">
    <table>
      <!---->
      <thead class="v-data-table__thead" role="rowgroup">
        <tr>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>Transaction</span>
              <!---->
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>Type</span>
              <!---->
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>Method</span>
              <!---->
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th--sorted v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>Mined On</span>
              <!---->
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>From</span>
              <!---->
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>To</span>
              <!---->
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>Value</span>
              <!---->
              <!---->
            </div>
          </th>
        </tr>
        <!---->
      </thead>
      <!---->
      <tbody class="v-data-table__tbody" role="rowgroup">
        <!---->
        <tr class="v-data-table__tr">
          <td class="v-data-table__td v-data-table-column--align-start">
            <hash-link-stub type="transaction" hash="0x123" fullhash="false" withname="true" notcopiable="false" withtokenname="false" xshash="true" unlink="false" loadcontract="false"></hash-link-stub>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start"><span class="v-chip v-chip--label v-theme--light text-primary v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false"><!----><span class="v-chip__underlay"></span>
            <!---->
            <!---->
            <div class="v-chip__content" data-no-activator="">CALL</div>
            <!---->
            <!----></span>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start"><span class="v-chip v-chip--label v-theme--light text-primary v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false" aria-describedby="v-tooltip-1"><!----><span class="v-chip__underlay"></span>
            <!---->
            <!---->
            <div class="v-chip__content" data-no-activator="">transfer</div>
            <!---->
            <!----></span>
            <!--teleport start-->
            <!--teleport end-->
          </td>
          <td class="v-data-table__td v-data-table-column--align-start">
            <div class="my-2 text-left">03/21 11:00:00 AM<br><small class="text-caption text-medium-emphasis">in 1 year</small></div>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start">
            <!--v-if-->
            <hash-link-stub type="address" hash="0xabc" fullhash="false" withname="true" notcopiable="false" withtokenname="true" xshash="false" unlink="false" loadcontract="false"></hash-link-stub>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start">
            <!--v-if-->
            <hash-link-stub contract="[object Object]" type="address" hash="0xdef" fullhash="false" withname="true" notcopiable="false" withtokenname="true" xshash="false" unlink="false" loadcontract="false"></hash-link-stub>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start">1 ETH</td>
        </tr>
        <!---->
        <!---->
      </tbody>
      <!---->
      <!---->
    </table>
  </div>
  <hr class="v-divider v-theme--light" aria-orientation="horizontal" role="separator">
  <div class="v-data-table-footer">
    <!---->
    <div class="v-data-table-footer__items-per-page"><span>Rows per page:</span>
      <div class="v-input v-input--horizontal v-input--center-affix v-input--density-compact v-theme--light v-locale--is-ltr v-input--dirty v-text-field v-select v-select--single v-select--selected">
        <!---->
        <div class="v-input__control">
          <div class="v-field v-field--active v-field--appended v-field--center-affix v-field--dirty v-field--no-label v-field--variant-outlined v-theme--light v-locale--is-ltr" role="combobox" aria-haspopup="listbox" aria-expanded="false" aria-controls="v-menu-5">
            <div class="v-field__overlay"></div>
            <div class="v-field__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background"></div>
                <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <div class="v-field__field" data-no-activator="">
              <!---->
              <!---->
              <!---->
              <div class="v-field__input" data-no-activator="">
                <!---->
                <!---->
                <div class="v-select__selection"><span class="v-select__selection-text">10<!----></span></div><input size="1" type="text" id="input-3" aria-describedby="input-3-messages" inputmode="none" aria-label="Open" title="Open" value="10">
              </div>
              <!---->
            </div>
            <!---->
            <div class="v-field__append-inner">
              <!----><i class="mdi-menu-down mdi v-icon notranslate v-theme--light v-icon--size-default v-select__menu-icon" aria-hidden="true"></i>
              <!---->
            </div>
            <div class="v-field__outline">
              <div class="v-field__outline__start"></div>
              <!---->
              <div class="v-field__outline__end"></div>
              <!---->
            </div>
          </div>
        </div>
        <!---->
        <!---->
      </div>
    </div>
    <div class="v-data-table-footer__info">
      <div>1-1 of 1</div>
    </div>
    <div class="v-data-table-footer__pagination">
      <nav class="v-pagination v-theme--light" role="navigation" aria-label="Pagination Navigation" data-test="v-pagination-root">
        <ul class="v-pagination__list">
          <li class="v-pagination__first" data-test="v-pagination-first"><button type="button" class="v-btn v-btn--disabled v-btn--icon v-theme--light v-btn--density-comfortable v-btn--rounded v-btn--size-default v-btn--variant-plain" disabled="" aria-label="First page" aria-disabled="true"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
              <!----><span class="v-btn__content" data-no-activator=""><!----></span>
              <!---->
              <!---->
            </button></li>
          <li class="v-pagination__prev" data-test="v-pagination-prev"><button type="button" class="v-btn v-btn--disabled v-btn--icon v-theme--light v-btn--density-comfortable v-btn--rounded v-btn--size-default v-btn--variant-plain" disabled="" aria-label="Previous page" aria-disabled="true"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
              <!----><span class="v-btn__content" data-no-activator=""><i class="mdi-chevron-left mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span>
              <!---->
              <!---->
            </button></li>
          <li class="v-pagination__next" data-test="v-pagination-next"><button type="button" class="v-btn v-btn--disabled v-btn--icon v-theme--light v-btn--density-comfortable v-btn--rounded v-btn--size-default v-btn--variant-plain" disabled="" aria-label="Next page" aria-disabled="true"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
              <!----><span class="v-btn__content" data-no-activator=""><i class="mdi-chevron-right mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span>
              <!---->
              <!---->
            </button></li>
          <li class="v-pagination__last" data-test="v-pagination-last"><button type="button" class="v-btn v-btn--disabled v-btn--icon v-theme--light v-btn--density-comfortable v-btn--rounded v-btn--size-default v-btn--variant-plain" disabled="" aria-label="Last page" aria-disabled="true"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
              <!----><span class="v-btn__content" data-no-activator=""><!----></span>
              <!---->
              <!---->
            </button></li>
        </ul>
      </nav>
    </div>
  </div>
</div>"
`;

exports[`TraceStepsTable.vue > Should render the table with highlighted address 1`] = `
"<div class="v-table v-table--has-top v-table--has-bottom v-theme--light v-table--density-default v-data-table hide-table-count" item-key="id">
  <!---->
  <div class="v-table__wrapper">
    <table>
      <!---->
      <thead class="v-data-table__thead" role="rowgroup">
        <tr>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>Transaction</span>
              <!---->
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>Type</span>
              <!---->
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>Method</span>
              <!---->
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th--sorted v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>Mined On</span>
              <!---->
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>From</span>
              <!---->
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>To</span>
              <!---->
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>Value</span>
              <!---->
              <!---->
            </div>
          </th>
        </tr>
        <!---->
      </thead>
      <!---->
      <tbody class="v-data-table__tbody" role="rowgroup">
        <!---->
        <tr class="v-data-table__tr">
          <td class="v-data-table__td v-data-table-column--align-start">
            <hash-link-stub type="transaction" hash="0x123" fullhash="false" withname="true" notcopiable="false" withtokenname="false" xshash="true" unlink="false" loadcontract="false"></hash-link-stub>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start"><span class="v-chip v-chip--label v-theme--light text-primary v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false"><!----><span class="v-chip__underlay"></span>
            <!---->
            <!---->
            <div class="v-chip__content" data-no-activator="">CALL</div>
            <!---->
            <!----></span>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start"><span class="v-chip v-chip--label v-theme--light text-primary v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false" aria-describedby="v-tooltip-1"><!----><span class="v-chip__underlay"></span>
            <!---->
            <!---->
            <div class="v-chip__content" data-no-activator="">transfer</div>
            <!---->
            <!----></span>
            <!--teleport start-->
            <!--teleport end-->
          </td>
          <td class="v-data-table__td v-data-table-column--align-start">
            <div class="my-2 text-left">03/21 11:00:00 AM<br><small class="text-caption text-medium-emphasis">in 1 year</small></div>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start"><span class="v-chip v-theme--light v-chip--density-default v-chip--size-x-small v-chip--variant-tonal mr-2" draggable="false"><!----><span class="v-chip__underlay"></span>
            <!---->
            <!---->
            <div class="v-chip__content" data-no-activator="">self</div>
            <!---->
            <!----></span>
            <hash-link-stub type="address" hash="0xabc" fullhash="false" withname="true" notcopiable="false" withtokenname="true" xshash="false" unlink="false" loadcontract="false"></hash-link-stub>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start">
            <!--v-if-->
            <hash-link-stub contract="[object Object]" type="address" hash="0xdef" fullhash="false" withname="true" notcopiable="false" withtokenname="true" xshash="false" unlink="false" loadcontract="false"></hash-link-stub>
          </td>
          <td class="v-data-table__td v-data-table-column--align-start">1 ETH</td>
        </tr>
        <!---->
        <!---->
      </tbody>
      <!---->
      <!---->
    </table>
  </div>
  <hr class="v-divider v-theme--light" aria-orientation="horizontal" role="separator">
  <div class="v-data-table-footer">
    <!---->
    <div class="v-data-table-footer__items-per-page"><span>Rows per page:</span>
      <div class="v-input v-input--horizontal v-input--center-affix v-input--density-compact v-theme--light v-locale--is-ltr v-input--dirty v-text-field v-select v-select--single v-select--selected">
        <!---->
        <div class="v-input__control">
          <div class="v-field v-field--active v-field--appended v-field--center-affix v-field--dirty v-field--no-label v-field--variant-outlined v-theme--light v-locale--is-ltr" role="combobox" aria-haspopup="listbox" aria-expanded="false" aria-controls="v-menu-6">
            <div class="v-field__overlay"></div>
            <div class="v-field__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background"></div>
                <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <div class="v-field__field" data-no-activator="">
              <!---->
              <!---->
              <!---->
              <div class="v-field__input" data-no-activator="">
                <!---->
                <!---->
                <div class="v-select__selection"><span class="v-select__selection-text">10<!----></span></div><input size="1" type="text" id="input-4" aria-describedby="input-4-messages" inputmode="none" aria-label="Open" title="Open" value="10">
              </div>
              <!---->
            </div>
            <!---->
            <div class="v-field__append-inner">
              <!----><i class="mdi-menu-down mdi v-icon notranslate v-theme--light v-icon--size-default v-select__menu-icon" aria-hidden="true"></i>
              <!---->
            </div>
            <div class="v-field__outline">
              <div class="v-field__outline__start"></div>
              <!---->
              <div class="v-field__outline__end"></div>
              <!---->
            </div>
          </div>
        </div>
        <!---->
        <!---->
      </div>
    </div>
    <div class="v-data-table-footer__info">
      <div>1-1 of 1</div>
    </div>
    <div class="v-data-table-footer__pagination">
      <nav class="v-pagination v-theme--light" role="navigation" aria-label="Pagination Navigation" data-test="v-pagination-root">
        <ul class="v-pagination__list">
          <li class="v-pagination__first" data-test="v-pagination-first"><button type="button" class="v-btn v-btn--disabled v-btn--icon v-theme--light v-btn--density-comfortable v-btn--rounded v-btn--size-default v-btn--variant-plain" disabled="" aria-label="First page" aria-disabled="true"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
              <!----><span class="v-btn__content" data-no-activator=""><!----></span>
              <!---->
              <!---->
            </button></li>
          <li class="v-pagination__prev" data-test="v-pagination-prev"><button type="button" class="v-btn v-btn--disabled v-btn--icon v-theme--light v-btn--density-comfortable v-btn--rounded v-btn--size-default v-btn--variant-plain" disabled="" aria-label="Previous page" aria-disabled="true"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
              <!----><span class="v-btn__content" data-no-activator=""><i class="mdi-chevron-left mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span>
              <!---->
              <!---->
            </button></li>
          <li class="v-pagination__next" data-test="v-pagination-next"><button type="button" class="v-btn v-btn--disabled v-btn--icon v-theme--light v-btn--density-comfortable v-btn--rounded v-btn--size-default v-btn--variant-plain" disabled="" aria-label="Next page" aria-disabled="true"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
              <!----><span class="v-btn__content" data-no-activator=""><i class="mdi-chevron-right mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span>
              <!---->
              <!---->
            </button></li>
          <li class="v-pagination__last" data-test="v-pagination-last"><button type="button" class="v-btn v-btn--disabled v-btn--icon v-theme--light v-btn--density-comfortable v-btn--rounded v-btn--size-default v-btn--variant-plain" disabled="" aria-label="Last page" aria-disabled="true"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
              <!----><span class="v-btn__content" data-no-activator=""><!----></span>
              <!---->
              <!---->
            </button></li>
        </ul>
      </nav>
    </div>
  </div>
</div>"
`;
